### ISSUE TITLE
Update solver in CTSM

### ISSUE THREAD

**Issue by ekluzek** [ESCOMP/CTSM#2422]
We need to update the decades old quadratic solver used in CTSM. 

@briandobbins

**Comment by ekluzek** [comment]
In ctsm5.1.dev173, a new Navy solver was brought in. We might as well use it for both FATES and non-FATES.

See here...

https://github.com/NGEET/fates/blob/main/main/FatesUtilsMod.F90

**Comment by ekluzek** [comment]
In discussion this morning, we thought a better solution would be to copy the code to the CESM share code.

If we have CTSM use FATES we end up with a double dependency in both directions which is bad. In this case it's a simple utility layer of FATES, so not as intertwined into FATES as other FATES layers.

The problem here is that fates_log() is being used as part of FatesGlobals, so we could NOT use it directly in CTSM.

**Comment by ekluzek** [comment]
@rgknox our plan right now is to copy it to share code and then have a CTSM update that points to the new share version with it in it. Let us know thoughts on that.

**Comment by rgknox** [comment]
hmmm, I'm sure there is some way to make this work. If we call these routines in FATES from the share, than this needs to be in the shared code for ELM as well.

**Comment by ekluzek** [comment]
Yes exactly. We weren't sure if they use the same share version or not. I'm guessing that ELM likely has their own version of share? But, we thought NORESMF might use the same one as CESM. Could you check?

**Comment by ekluzek** [comment]
In order to get this to work so that we'd share it from FATES, we'd need to do something like have a directory under FATES for a utility layer that doesn't allow use of ANY other FATES code. So for example here, the use of fates_log() would be replaced by passing the log unit into the subroutine. The rule for that directory would be that anything in that directory is completely independent of FATES.

We do always build with FATES -- we just don't always initialize FATES -- so we can't be using higher level FATES subroutines and objects in non-FATES CTSM. So we'd need a directory that's independent.

The other way to do it would be to have a new share external that all FATES models would use. That would also work, but would be adding in something new, that would take a bit of organizing around.

@adrifoster what do you think about this idea?

**Comment by mvdebolskiy** [comment]
@rgknox  we are checking out a [fork](https://github.com/NorESMHub/NorESM_share) of the CESM_share in NORESM. So there is no problem to pull from the upstream for us.

**Comment by ekluzek** [comment]
Bringing this up again, as we want to make sure the update won't change answers. If it does change answers we'll want it in right away.

**Comment by wwieder** [comment]
I know we don't like duplicated code, but it seems like efforts to share code may have bogged us down.  Is there an expedient option we can implement  in CTSM without too much trouble?

**Comment by samsrabin** [comment]
Per SE meeting today, this _doesn't_ change answers.
